hotfix/188 fix put and patch donation

JuanluRM opened the pull request:
Changed permission of Donations for partner roles. Now is able to modify (put/patch) or cancel (delete) his donation plan as intended.

auroranavas reviewed (changes requested):
Partners can modify their donations well, however, the IBAN can be changed. The issue says that the IBAN can't be modified. Were you not able to find out how?

JuanluRM commented:
I added the validation that iban cannot be changed in put or patch. I also added that lesson can be updated through patch since front end needs the method.

auroranavas reviewed (changes requested):
The IBAN validator works well, both with PATCH and PUT. About the other feature on Lesson, you need to add support for the PATCH in the validator in LessonSerializer. This is the error I got when I patched the students:
```
TypeError at /api/lesson/1/
'<' not supported between instances of 'NoneType' and 'int'
```

JuanluRM commented:
The problem is when input patch some fields are null and the validators can't identified them and thus error appear. It should be fixed now.

auroranavas commented on nexong/api/Lesson/lessonSerializer.py:
```diff
@@ -30,17 +30,18 @@ def validate(self, attrs):
-            validation_error[
-                "capacity"
-            ] = "capacity must be higher or equal to the number of attendees selected."
+        if max_attendees is not None:
+            if max_attendees < 1:
+                validation_error["capacity"] = "capacity must be higher than 0."
+            attendees = attrs.get("students")
+            if attendees:
+                num_attendees = len(attendees)
+            else:
+                num_attendees = 0
+            if max_attendees < num_attendees:
+                validation_error[
+                    "capacity"
+                ] = "capacity must be higher or equal to the number of attendees selected."
```
This does not exactly work. I patched the capacity field with 1, but the lesson already contains 2 students and an error is not raised. The reason is that if capacity is patched, only capacity is sent to backend, so, `attrs.get('students')` is null.

auroranavas reviewed (changes requested):
Read the comment I left in your code. To mitigate the error, I see two options:

1. Find a way for this to work with a PATCH operation. The problem is that the validation for this class depends on a few attributes, so the PATCH is problematic. Maybe there are ways to handle this in Django.
2. Talk with frontend developers to leave the PUT operation only.

Maybe you should open a different issue for this. Frontend probably needs the donation change, which you already solved.

JuanluRM commented:
I fully implemented the patch method for the lesson class and added extra permissions that front end requested. Everything should run correctly now.

auroranavas reviewed (changes requested):
It seems that now there is a problem with the start_date of a Lesson when I PATCH some other attribute. I tried to PATCH the capacity of a Lesson but I get this error: `"start_date": [
        "The start date must be in the future."
    ]`

JuanluRM commented:
I believe that's working as intended because now it always checks the necesary validations even if they are not included in the PATCH, is made this way to avoid for example cases where an user PATCH an end_date but not a start_date it may skip the validation. The error you are seeing is probably becasue the object has the dates wrong already from the populate.json and when trying to modify it, it goes through the validations again. If that's not the case and the object has the correct fields should pass, report it again.

auroranavas reviewed (approved):
Thank you for listening to my feedback. You did a good job!
